src/IO/StdStreamBufFromReadBuffer.cpp

```diff
@@ -0,0 +1,111 @@
+#include <IO/StdStreamBufFromReadBuffer.h>
+#include <IO/SeekableReadBuffer.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int SEEK_POSITION_OUT_OF_BOUND;
+}
+
+
+StdStreamBufFromReadBuffer::StdStreamBufFromReadBuffer(std::unique_ptr<ReadBuffer> read_buffer_)
+{
+    read_buffer = std::move(read_buffer_);
+    seekable_read_buffer = dynamic_cast<SeekableReadBuffer *>(read_buffer.get());
+}
+
+StdStreamBufFromReadBuffer::StdStreamBufFromReadBuffer(ReadBuffer & read_buffer_)
+{
+    if (dynamic_cast<SeekableReadBuffer *>(&read_buffer_))
+    {
+        read_buffer = wrapSeekableReadBufferReference(static_cast<SeekableReadBuffer &>(read_buffer_));
+        seekable_read_buffer = static_cast<SeekableReadBuffer *>(read_buffer.get());
+    }
+    else
+    {
+        read_buffer = wrapReadBufferReference(read_buffer_);
+    }
+}
+
+StdStreamBufFromReadBuffer::~StdStreamBufFromReadBuffer() = default;
+
+int StdStreamBufFromReadBuffer::underflow()
+{
+    char c;
+    if (!read_buffer->peek(c))
+        return std::char_traits<char>::eof();
+    return c;
+}
+
+std::streamsize StdStreamBufFromReadBuffer::showmanyc()
+{
+    return read_buffer->available();
+}
+
+std::streamsize StdStreamBufFromReadBuffer::xsgetn(char_type* s, std::streamsize count)
+{
+    return read_buffer->read(s, count);
+}
+
+std::streampos StdStreamBufFromReadBuffer::seekoff(std::streamoff off, std::ios_base::seekdir dir, std::ios_base::openmode which)
+{
+    if (dir == std::ios_base::beg)
+        return seekpos(off, which);
+
+    if (dir == std::ios_base::cur)
+        return seekpos(getCurrentPosition() + off, which);
+
+    throw Exception(ErrorCodes::LOGICAL_ERROR, "Wrong seek's base {}", static_cast<int>(dir));
```

Comment:
Interesting. There is should be case for std::ios_base::end also.
I definitely have seen that S3 do '.seekg(0, body->end);'.
When it sets context-length in httpRequest (actually that seems like a bug because that information has been provided).

Reply:
The problem is that `ReadBuffer` doesn't have an interface to get the total size. However it's not a problem, I can pass it via a new parameter of the constructor of `StdStreamBufFromReadBuffer`.

Reply:
done